File: python/dfs.py
import unittest
from graph import *

def dfs(graph, start):
    stack = [start]
    traversed = []

    while stack:
        curr = stack.pop()
        if not curr.traversed:
            curr.traversed = True
            traversed.append(curr)
            stack.extend([v for v in curr.neighbors])

    return traversed

def dfs_recursive(graph, curr):
    traversed = []
    def dfs_helper(curr):
        if not curr.traversed:
            curr.traversed = True
            traversed.append(curr)
            for neighbor in curr.neighbors:
                dfs_helper(neighbor)
    dfs_helper(curr)
    return traversed

class TestDFS(unittest.TestCase):
    def test_one_edge(self):
        v1 = Vertex("me")
        v2 = Vertex("you")
        g = Graph({v1, v2})
        g.add_edge(v1, v2)
        path = dfs(g, v1)
        self.assertTrue(path in [
            [v1, v2]
        ])

        v1 = Vertex("me")
        v2 = Vertex("you")
        g = Graph({v1, v2})
        g.add_edge(v1, v2)
        path = dfs_recursive(g, v1)
        self.assertTrue(path in [
            [v1, v2]
        ])

    def test_few(self):
        d = Vertex("d")
        c = Vertex("c", {d})
        b = Vertex("b", {c})
        a = Vertex("a", {b, c, d})
        g = Graph({a, b, c, d})

        path = dfs(g, a)
        self.assertTrue(path in [
            [a, b, c, d],
            [a, d, b, c],
            [a, c, d, b],
            [a, d, c, b]
        ])

        d = Vertex("d")
        c = Vertex("c", {d})
        b = Vertex("b", {c})
        a = Vertex("a", {b, c, d})
        g = Graph({a, b, c, d})

        path = dfs_recursive(g, a)
        self.assertTrue(path in [
            [a, b, c, d],
            [a, d, b, c],
            [a, c, d, b],
            [a, d, c, b]
        ])


if __name__ == "__main__":
    unittest.main()

File: javascript/dfs.js
'use strict';

const GraphNode = require('./Graph').GraphNode;

function dfs(graph, v1, v2) {
  let stack = [v1];

  while (stack.length > 0) {
    const curr = stack.pop();

    if (curr.val === v2.val) {
      curr.visited = true;
      return curr;
    }

    if (!(curr.visited)) {
      curr.visited = true;
      stack.push(...graph[curr.val]);
    }
  }
}

function dfsRecursive(graph, v1, v2) {
  v1.visited = true;

  if (v1.val === v2.val) {
    return v2;
  }

  const neighbors = graph[v1.val];
  for (let i = 0; i < neighbors.length; i++) {
    const n = neighbors[i];
    if (!n.visited) {
      let found = dfsRecursive(graph, n, v2);
      if (found) {
        return found;
      }
    }
  }

  return null;
}

function test() {
  const a = new GraphNode('a');
  const b = new GraphNode('b');
  const c = new GraphNode('c');
  const d = new GraphNode('d');
  const e = new GraphNode('e');
  const f = new GraphNode('f');
  let nodes = [a, b, c, d, e, f];

  // Represent graph as an adjacency list
  const graph = {
    a: [b, c],
    b: [d],
    c: [d, e, f],
    d: [e],
    e: [],
    f: []
  };

  const result = dfs(graph, a, f);
  console.log('result: ', result);

  // Reset
  nodes = nodes.map(n => n.visited = false);

  const resultRecursive = dfsRecursive(graph, a, f);
  console.log('recursive: ', resultRecursive);
}

test();